### PR TITLE
Allow for ACCELERATE_SEED env var

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR introduces `ACCELERATE_SEED` as part of `set_seed`. Doing so will allow for the custom random samplers to rely on it for setting the seed, rather than the generator (as this can lead to imperfect issues). Solves related trainer test failures.

Fixes # (issue)

[Internal slack thread](https://huggingface.slack.com/archives/C01NE71C4F7/p1699294942410279)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik @BenjaminBossan

### CODE REVIEW

File: docs/source/package_reference/utilities.md [hunk 1]
```diff
@@ -50,6 +50,13 @@ These are basic dataclasses used throughout 🤗 Accelerate and they can be pass
 
 [[autodoc]] utils.ProjectConfiguration
 
+## Environmental Variables
+
+These are environmental variables that can be enabled for different use cases
+
+* `ACCELERATE_SEED` (`int`): The seed to use for different internal random states not covered through the base cases. Also set during [`utils.set_seed`]
```
Comment (BenjaminBossan) [inline docs/source/package_reference/utilities.md:57]: The description is a bit vague. As a user, what would interest me is in what circumstances I have to set this env var for reproducibility.

File: docs/source/package_reference/utilities.md [hunk 2]
```diff
@@ -50,6 +50,13 @@ These are basic dataclasses used throughout 🤗 Accelerate and they can be pass
 
 [[autodoc]] utils.ProjectConfiguration
 
+## Environmental Variables
```
Comment (BenjaminBossan) [inline docs/source/package_reference/utilities.md:53]: Good idea to document those in a separate section.

File: src/accelerate/utils/random.py [hunk 3]
```diff
@@ -30,7 +31,8 @@
 
 def set_seed(seed: int, device_specific: bool = False):
     """
-    Helper function for reproducible behavior to set the seed in `random`, `numpy`, `torch`.
+    Helper function for reproducible behavior to set the seed in `random`, `numpy`, `torch`. Also will set the random
+    seed for `SeedableRandomSampler` through the `ACCELERATE_SEED` environment variable.
```
Comment (BenjaminBossan) [inline src/accelerate/utils/random.py:35]: Similar comment to the one above: Users don't really care about `SeedableRandomSampler`, as this is an implementation detail of accelerate. They will care more about knowing when this function needs to be used.

**Review by BenjaminBossan — APPROVED** [review]
Implementation-wise, this looks good to me. I think the documentation could be improved, LMK if you agree.

File: docs/source/package_reference/utilities.md (hunk 1 above)
Comment (muellerzr) [reply docs/source/package_reference/utilities.md]: Went with new implementation, irrelevent :)

**Review by BenjaminBossan — APPROVED** [review]
Nice and simple. So in the end, users should call `set_seed` to ensure reproducibility and that's enough, right?

**Comment by muellerzr** [comment]
@BenjaminBossan yep! set_seed, and in the Trainer just having it set the torch seed will be enough for general users

**Pull request merged**